tuna/miopen/celery_tuning/tuning.py

```diff
@@ -0,0 +1,395 @@
+
+  try:
+    worker_type, kwargs, fdb_attr, q_name = prep_tuning(library)
+  except ValueError as verr:
+    LOGGER.error(verr)
+    return False
+
+  #if enqueue_only is False, we only launch the workers
+  if not library.args.enqueue_only:
+    return True
+
+  res_set = ResultSet([])
+  start = time.time()
+
+  with DbSession() as session:
+    while True:
+      try:
+        job_list = []
+        #get all the jobs from mySQL
+        job_list = library.get_jobs(session, library.fetch_state,
+                                    library.set_state, library.args.session_id,
+                                    job_batch_size)
+
+        for i in range(0, len(job_list), job_batch_size):
+          batch_jobs = job_list[i:min(i + job_batch_size, len(job_list))]
+          if library.args.fin_steps:
+            entries = library.compose_work_objs_fin(session, batch_jobs,
+                                                    library.dbt)
+
+          entries = library.compose_work_objs_fin(session, job_list,
+                                                  library.dbt)
+          serialized_jobs = serialize_chunk(entries)
+
+          #build context for each celery task
+          for job, config in serialized_jobs:
+            context = {
+                'job': job,
+                'config': config,
+                'worker_type': worker_type,
+                'arch': library.dbt.session.arch,
+                'num_cu': library.dbt.session.num_cu,
+                'kwargs': kwargs,
+                'fdb_attr': fdb_attr
+            }
+
+            #calling celery task, enqueuing to celery queue
+            res_set.add(celery_enqueue.apply_async((context,), queue=q_name))
+
+        if not job_list:
+          if not res_set:
+            return False
+          LOGGER.info('All tasks added to queue')
+          break
+      except KeyboardInterrupt:
+        LOGGER.error('Keyboard interrupt caught, draining results queue')
+        session.rollback()
+        stop_active_workers()
+        results_gather(res_set, worker_type)
+        purge_queue([q_name], LOGGER)
+
+  results_gather(res_set, worker_type)
+  stop_active_workers()
+  end = time.time()
+  LOGGER.info("Took {:0>8}".format(str(timedelta(seconds=end - start))))  #pylint: disable=consider-using-f-string
+
+  return True
+
+
+def results_gather(res_set, worker_type):
+  """Start subproc to drain result queue and populate mysql DB"""
+  LOGGER.info('Started drain process')
+  #start draining result_queue in subprocess
+  drain_process = Process(target=drain_queue, args=[worker_type, DBT])
+  drain_process.start()
+
+  LOGGER.info('Gathering async results in callback function, blocking')
+  #waiting on all results to come in
+  _ = res_set.join(callback=result_callback)
+
+  #terminate result_queue drain process with special queue token (NONE,NONE)
+  LOGGER.info('Adding terminatig token to result_queue')
+  result_queue.put([None, None])
+
+  drain_process.join()
```

Comment:
Do you monitor the number of `[None, None]` tokens to determine if all threads have finished? I believe you mentioned that you just monitor if there are any unfinshed jobs left, but just to confirm again...

Reply:
this is in the main process so there is only 1 token of this kind being added to the Q. the minute we see it, we terminate waiting on results because all the results have already come in

Reply:
The `results_gather` function first launches drain_queue, and then puts [None, None] onto the queue. Is it possible that the queue might get drained even before a [None, None] is put there?

Reply:
I'm not really familiar with multithreading, but does results_gather launch multiple drain_queue processes in parallel? Because in [tuning.py#L361](https://github.com/ROCm/MITuna/blob/2751d98346e1789823b4a013517d12ae707c8c7c/tuna/miopen/celery_tuning/tuning.py#L361), drain_queue uses the `get` function to remove an element off the queue and it then checks if the element is [None, None]. If it is, it returns. But since [None, None] has now been removed from the queue, the other drain_queue threads would will not hit the terminating condition.

Reply:
there is only 1 drain_queue thread (this part is not multithreaded), the results Q that we drain is not shared by multiple processes. The none, none token is added to the results Q, after we do a thread join on the the threads that run the actual work and that works gets added to the results Q. Since join is blocking, [none, none] gets added at the end of the queue after all results have come in. When we dequeue and see the token that means we have reached the end of the Q.